File: apps/ios/Pebbles/Features/Path/Read/BannerAspect.swift
import CoreGraphics

/// Banner aspect-ratio bucket chosen for a source image. The pebble read
/// banner snaps the source's width/height ratio to the nearest of four fixed
/// buckets — 16:9, 4:3, 1:1, 3:4 — so landscape, square, and portrait uploads
/// are each shown at a sensible ratio (issue #599 added the 3:4 portrait
/// bucket).
///
/// Pure value type. No view dependencies; trivially unit-testable.
enum BannerAspect: Equatable {
    case sixteenNine
    case fourThree
    case square
    case threeFour

    /// CG ratio (width / height) for the bucket.
    var cgRatio: CGFloat {
        switch self {
        case .sixteenNine: return 16.0 / 9.0
        case .fourThree:   return 4.0 / 3.0
        case .square:      return 1.0
        case .threeFour:   return 3.0 / 4.0
        }
    }

    /// Pick the bucket whose `cgRatio` is closest to `ratio` (absolute
    /// distance). Portrait sources (`ratio < 1`) now bucket to `.threeFour`
    /// (0.75) rather than collapsing to `.square`.
    static func nearest(to ratio: CGFloat) -> BannerAspect {
        let candidates: [BannerAspect] = [.sixteenNine, .fourThree, .square, .threeFour]
        return candidates.min(by: { abs($0.cgRatio - ratio) < abs($1.cgRatio - ratio) })
            ?? .sixteenNine
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/read/BannerAspect.kt
package app.pbbls.android.features.path.read

import kotlin.math.abs

/**
 * Aspect-ratio bucket chosen for a pebble's snap on the read view — ports iOS
 * `BannerAspect.swift`. The snap's width/height ratio snaps to the nearest of
 * three fixed buckets — 4:3 (landscape), 1:1, 3:4 (portrait) — so the imported
 * photo reads at a natural shape instead of a forced strip (issue #599).
 *
 * These buckets replace the earlier 16:9/4:3/1:1 set: the redesigned page
 * frames the whole picture (portrait included) with the Petroglyph overlapping
 * its top-right, so a tall 3:4 crop is now first-class and the wide 16:9 strip
 * is gone.
 *
 * Pure value type; trivially unit-testable.
 */
enum class BannerAspect(
    /** Width / height for the bucket. */
    val ratio: Float,
) {
    FOUR_THREE(4f / 3f),
    SQUARE(1f),
    THREE_FOUR(3f / 4f),
    ;

    companion object {
        /**
         * The bucket whose ratio is closest to [ratio] (absolute distance).
         * Very wide sources collapse to [FOUR_THREE] and very tall ones to
         * [THREE_FOUR] since those are the extreme candidates — no special case.
         */
        fun nearest(ratio: Float): BannerAspect = entries.minByOrNull { abs(it.ratio - ratio) } ?: SQUARE
    }
}
